router/core/executor.go

```diff
@@ -72,32 +72,43 @@ func (b *ExecutorConfigurationBuilder) Build(ctx context.Context, routerConfig *
 	// this is the resolver, it's stateful and manages all the client connections, etc...
 	resolver := resolve.New(ctx, options)
 
-	// this is the GraphQL Schema that we will expose from our API
-	var definition ast.Document
-	var report operationreport.Report
-	// The client schema may not be present in old configs
-	if routerConfig.EngineConfig.GetGraphqlClientSchema() != "" {
-		definition, report = astparser.ParseGraphqlDocumentString(routerConfig.EngineConfig.GetGraphqlClientSchema())
-	} else {
-		definition, report = astparser.ParseGraphqlDocumentString(routerConfig.EngineConfig.GraphqlSchema)
-	}
+	routerSchemaDefinition, report := astparser.ParseGraphqlDocumentString(routerConfig.EngineConfig.GraphqlSchema)
 	if report.HasErrors() {
 		return nil, fmt.Errorf("failed to parse graphql schema from engine config: %w", report)
 	}
-
-	// we need to merge the base schema, it contains the __schema and __type queries
-	// these are not usually part of a regular GraphQL schema
+	// we need to merge the base schema, it contains the __schema and __type queries,
+	// as well as built-in scalars like Int, String, etc...
+	// these are usually not part of a regular GraphQL schema
 	// the engine needs to have them defined, otherwise it cannot resolve such fields
-	err = asttransform.MergeDefinitionWithBaseSchema(&definition)
+	err = asttransform.MergeDefinitionWithBaseSchema(&routerSchemaDefinition)
 	if err != nil {
 		return nil, fmt.Errorf("failed to merge graphql schema with base schema: %w", err)
 	}
 
+	// this is the GraphQL Schema that we will expose from our API
+	// it should be used for the introspection and query validation
+	var clientSchemaDefinition *ast.Document
+
+	if clientSchemaStr := routerConfig.EngineConfig.GetGraphqlClientSchema(); clientSchemaStr != "" {
+		clientSchema, report := astparser.ParseGraphqlDocumentString(clientSchemaStr)
+		if report.HasErrors() {
+			return nil, fmt.Errorf("failed to parse graphql client schema from engine config: %w", report)
+		}
+		err = asttransform.MergeDefinitionWithBaseSchema(&clientSchema)
+		if err != nil {
+			return nil, fmt.Errorf("failed to merge graphql client schema with base schema: %w", err)
+		}
+		clientSchemaDefinition = &clientSchema
+	} else {
+		clientSchemaDefinition = &routerSchemaDefinition
```

Comment:
This is a backwards compatible fallback, right? Can we mark it as such and put a date for removal?

Reply:
yes

What should be a timeframe for this?

Should we log warning? e.g. "you are using outdated wgc version"

Reply:
No, the ClientSchema is optional and only used with schema contracts. If schema contracts aren't used the router schema is the client and router schema.

Reply:
not fully true

In case schema has `inaccessible` directive there will be 2 schemas

Reply:
The client schema is a product of `@inaccessible` AND/OR `@tag` directives.

Reply:
Yes, but still it is nothing we want to get rid of in the nearly future.